modules/dnn/src/dnn.cpp

```diff
@@ -430,19 +430,19 @@ struct DataLayer : public Layer
                backendId == DNN_BACKEND_INFERENCE_ENGINE && inputsData.size() == 1;
     }
 
-    void forward(InputArrayOfArrays inputs, OutputArrayOfArrays outputs, OutputArrayOfArrays internals) CV_OVERRIDE
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
     {
         CV_TRACE_FUNCTION();
         CV_TRACE_ARG_VALUE(name, "name", name.c_str());
 
         CV_OCL_RUN(IS_DNN_OPENCL_TARGET(preferableTarget),
-                   forward_ocl(inputs, outputs, internals));
+                   forward_ocl(inputs_arr, outputs_arr, internals_arr) ||
+                   forward_fallback(inputs_arr, outputs_arr, internals_arr));
```

Comment:
`CV_OCL_RUN()` should contain OpenCL-related calls only.
It is just a "hook" for OpenCL processing.

Why is `forward_fallback` here? What is the real purpose of "fallback" call?

Reply:
`forward_fallback` calls Layer's forward method converting FP16 mats to FP32 and back to FP16 after `forward`. For now, `forward_fallback` calls forward with `std::vector<Mat*>` but in this PR we remove it.

In this PR an every forward looks like
```cpp
void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
{
    CV_TRACE_FUNCTION();
    CV_TRACE_ARG_VALUE(name, "name", name.c_str());

    CV_OCL_RUN(IS_DNN_OPENCL_TARGET(preferableTarget) &&
               forward_ocl(inputs_arr, outputs_arr, internals_arr))

    if (inputs_arr.depth() == CV_16S)
    {
        forward_fallback(inputs_arr, outputs_arr, internals_arr);
        return;
    }

    // Get Mats from input arrays and perform default CPU implementation
    // ...
}
```